metashare/api/jobs.py

```diff
@@ -293,3 +297,33 @@ def refresh_github_repositories_for_user(user):
 
 
 refresh_github_repositories_for_user_job = job(refresh_github_repositories_for_user)
+
+
+# This avoids partially-applied saving:
+@transaction.atomic
+def refresh_commits(*, repository, branch_name):
+    """
+    This should only run when we're notified of a force-commit. It's the
+    nuclear option.
+    """
+    from .models import Task
+
+    user = repository.get_a_matching_user()
+    if user is None:
+        logger.warning(f"No matching user for repository {repository.pk}")
+        return
+    repo = get_repo_info(user, repository.repo_id)
+    # We get this as a GitHubIterator, but we want to slice it later, so
+    # we will convert it to a list:
+    commits = list(repo.commits(repo.branch(branch_name).latest_sha()))
```

Comment:
I don't like this approach because it will make unnecessary requests to retrieve all commits in the branch history, not just the ones since the origin_sha. It would be better to iterate over the iterator returned by repo.commits() in a for loop, adding commits to a list as you go and breaking when you get to the origin_sha.

Reply:
Ah, I see that we're potentially updating multiple tasks with different origins. So I guess that would mean not breaking out of the loop until we've found the origin_sha for all the tasks.

But I'm not sure I understand the scenario this is trying to handle. The situation where we would have multiple tasks on the same branch is if the tasks are still pointing at the project branch because no dev org has been created yet, right? In that case if the project branch is reset then I think it would be more appropriate to reset the origin_sha on the tasks and leave the task commit lists empty so the tasks still show no changes compared to the project.

Reply:
@davisagli I think you're generally correct, in that there is no use-case for there being multiple tasks. This doesn't run at all for tasks that have no branch yet, so it should always find zero or one task for the giving `branch_name`.

Reply:
We expect it to find zero or one, but I wanted it to be able to run coherently and not crash on >1, as there are no actual protections against >1 being present (and as things stand, that would take more than a `unique_together`, I think, to achieve).

Reply:
In order to handle >1, it needs to get all the commits first. Until we know that that is an excessive API call to handle in a worker process, or that it strains memory, I'd consider changing it to be premature optimization.

Reply:
I'm concerned about the number of github API requests it will use if we set up MetaShare for a repository with a large commit history.

Reply:
We could also set `number=1000` or something and presume that the original SHA is within that range, as a quick approximation. That would be a reasonable minimal change.